### PR TITLE
Api create image

### PR DESCRIPTION
**Pull request by SomewhatDamaged**
### Before you begin

* I understand my contributions may be rejected for any reason
* I understand my contributions are for the benefit of Derpibooru and/or the Philomena software
* I understand my contributions are licensed under the GNU AGPLv3

- [x] I understand all of the above

---

Adds the `POST` method to `/api/v1/json/images` to allow for API-based creation of images.

Requires API key be used. Logs IP address. Sets fingerprint to `c415049`.

Same tag requirements as site uploading.
Description and source are optional.

Will return newly created image and a `200` on success.
Will return nothing and `401` if API key missing or invalid.
Will return nothing and `403` if IP is banned or API key exists and user is banned (as per normal banning rules, excluding `fingerprint`).
Will return errors and a `400` if image could not be created.

### CODE REVIEW

File: lib/philomena_web/controllers/api/json/image_controller.ex [hunk 1]
```diff
@@ -27,4 +35,72 @@ defmodule PhilomenaWeb.Api.Json.ImageController do
...
+      |> Map.merge(image_params)
+
+    case user do
+      nil ->
+        conn
+        |> put_status(:unauthorized)
+        |> text("")
+
+      _ ->
+        attributes = %{
```
Comment (liamwhite) [inline lib/philomena_web/controllers/api/json/image_controller.ex:57]: Use `UserAttributionPlug` to get the attributes, and just override the fingerprint

File: lib/philomena_web/controllers/api/json/image_controller.ex [hunk 2]
```diff
@@ -27,4 +35,72 @@ defmodule PhilomenaWeb.Api.Json.ImageController do
         render(conn, "show.json", image: image, interactions: interactions)
     end
   end
+
+  def create(conn, %{"image" => image_params} = params) do
+    user = conn.assigns.current_user
+
+    params =
```
Comment (liamwhite) [inline lib/philomena_web/controllers/api/json/image_controller.ex:42]: Why is this necessary? Why can't you just use `image_params`?

File: lib/philomena_web/controllers/api/json/image_controller.ex [hunk 3]
```diff
@@ -27,4 +35,72 @@ defmodule PhilomenaWeb.Api.Json.ImageController do
...
+      params
+      |> Map.put("scraper_url", Map.get(params, "url"))
+      |> Map.put("anonymous", Map.get(params, "anonymous", false))
+      |> Map.put("source_url", Map.get(params, "source_url", ""))
+      |> Map.put("description", Map.get(params, "description", ""))
+      |> Map.merge(image_params)
+
+    case user do
```
Comment (liamwhite) [inline lib/philomena_web/controllers/api/json/image_controller.ex:50]: Not a knock against your implementation, but I think this would be better written with its own plug (`ApiRequireAuthorizedPlug`).

File: lib/philomena_web/controllers/api/json/image_controller.ex [hunk 4]
```diff
@@ -27,4 +35,72 @@ defmodule PhilomenaWeb.Api.Json.ImageController do
...
+            spawn(fn ->
+              Images.repair_image(image)
+            end)
+
+            # ImageProcessor.cast(image.id)
+            Images.reindex_image(image)
+            Tags.reindex_tags(image.added_tags)
+            UserStatistics.inc_stat(user, :uploads)
+
+            interactions = Interactions.user_interactions([image], user)
```
Comment (liamwhite) [inline lib/philomena_web/controllers/api/json/image_controller.ex:74]: Interactions is always empty on creation, so it doesn't actually need to be loaded; just pass an empty list.

File: lib/philomena_web/controllers/api/json/image_controller.ex (hunk 2 above)
Comment (SomewhatDamaged) [reply lib/philomena_web/controllers/api/json/image_controller.ex:42]: Yeah. Worked it better so it uses `image_params` and only adds the things from `params` that are strictly required to make the image (metadata, `scraper_url`, `anonymous`).

Comment (SomewhatDamaged) [reply lib/philomena_web/controllers/api/json/image_controller.ex]: This should be much neater now.
